### PR TITLE
Refactor Scene Objects

### LINKED ISSUE #46: Refactor `DataType` and `Shade`
`DataType` feels too general. Maybe `ShaderData`?

_Originally posted by @Licini in https://github.com/compas-dev/compas_viewer/pull/40#discussion_r1444416929_
            

- change the general names `_read_points_data`, `DataType`, etc.
- `numpy` then for better performance.

### PR DESCRIPTION
**Pull request by ZacZhangzhuo**
Followed by the discussion with @tomvanmele on 6th March, there is the refactor solution of the viewer.

## Major Changes
-   Fully inherent `compas.scene.MeshObject` and `compas.scene.GeometryObject`.
-   Added `ViewerGeometryObject` as the abstract class for all the geometry objects. Other specific geometry objects are inherited from this class.
-   Unify colour naming. variables that control the colours of geometries are `surfacecolor`, `linecolor`, and `pointcolor`, yet variables that control the colours of meshes are `facecolor`, `edgecolor`, and `vertexcolor`.
-   Rename typing allies, resolve #46 .

Updated example files will follow.

### CODE REVIEW

File: src/compas_viewer/scene/geometryobject.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+from compas.geometry import Geometry
+from compas.geometry import Line
+from compas.geometry import Point
+from compas.scene import GeometryObject as BaseGeometryObject
+
+from .sceneobject import ShaderDataType
+from .sceneobject import ViewerSceneObject
+
+
+class GeometryObject(ViewerSceneObject, BaseGeometryObject):
```
Comment (tomvanmele) [inline src/compas_viewer/scene/geometryobject.py:14]: it might make sense to differentiate geometry objects further (for example into curves, surfaces, and shapes), because this might make conversion of parametric geometry into discrete geometry more efficient, but we can do that at a later stage...

File: src/compas_viewer/scene/geometryobject.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    @property
+    def lines(self) -> Optional[list[Line]]:
+        """The lines to be shown in the viewer."""
+        raise NotImplementedError
+
+    @property
+    def surfaces(self) -> Optional[list[Tuple[Point, Point, Point]]]:
+        """The surface to be shown in the viewer. Currently only triangles are supported."""
+        raise NotImplementedError
+
```
Comment (tomvanmele) [inline src/compas_viewer/scene/geometryobject.py:83]: perhaps it would make sense to add a "viewmesh" property?

File: src/compas_viewer/scene/graphobject.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
+from compas.geometry import Point
 from compas.scene import GraphObject as BaseGraphObject
 
-from .sceneobject import DataType
-from .sceneobject import ViewerSceneObject
+from .geometryobject import GeometryObject as ViewerGeometryObject
 
 
-class GraphObject(ViewerSceneObject, BaseGraphObject):
+class GraphObject(ViewerGeometryObject, BaseGraphObject):
```
Comment (tomvanmele) [inline src/compas_viewer/scene/graphobject.py:12]: a graph object is not a geometry object.

File: src/compas_viewer/scene/lineobject.py [hunk 4]
```diff
@@ -14,24 +17,21 @@ class LineObject(ViewerSceneObject, GeometryObject):
...
-        colors = [
-            self.linescolor.get(0, self.linescolor["_default"]),  # type: ignore
-            self.linescolor.get(1, self.linescolor["_default"]),  # type: ignore
-        ]
-        elements = [[0, 1]]
-        return positions, colors, elements
+    @property
+    def surfaces(self) -> Optional[list[Tuple[Point, Point, Point]]]:
+        """The surface to be shown in the viewer. Currently only triangles are supported."""
+        pass
```
Comment (tomvanmele) [inline src/compas_viewer/scene/lineobject.py:37]: if the behaviour is based on an implicit `None` return, i would explicitly return `None`, instead of passing...

File: src/compas_viewer/scene/graphobject.py [hunk 5]
```diff
@@ -26,31 +30,17 @@ def __init__(self, graph: Graph, **kwargs):
...
+
+    @property
+    def lines(self) -> Optional[list[Line]]:
+        """The lines to be shown in the viewer."""
+        return [Line(self.graph.node_coordinates(u), self.graph.node_coordinates(v)) for u, v in self.graph.edges()]
+
+    @property
+    def surfaces(self) -> Optional[list[Tuple[Point, Point, Point]]]:
+        """The surface to be shown in the viewer. Currently only triangles are supported."""
+        pass
```
Comment (tomvanmele) [inline src/compas_viewer/scene/graphobject.py:46]: if the behaviour is based on an implicit None return, i would explicitly return None, instead of passing...

File: src/compas_viewer/scene/meshobject.py [hunk 6]
```diff
@@ -18,19 +23,23 @@ class MeshObject(ViewerSceneObject, BaseMeshObject):
     ----------
     mesh : :class:`compas.datastructures.Mesh`
         A COMPAS mesh.
+    vertexcolor : Union[Dict[Any, :class:`compas.colors.Color`], :class:`compas.colors.Color`]], optional
+        The vertex color. Global settings in the viewer will be used if not specified.
```
Comment (tomvanmele) [inline src/compas_viewer/scene/meshobject.py:27]: i would rephrase this everywhere to "Default is the value of `pointcolor` in `viewer.config`."

File: src/compas_viewer/scene/meshobject.py [hunk 7]
```diff
@@ -42,42 +51,68 @@ class MeshObject(ViewerSceneObject, BaseMeshObject):
...
         positions = []
         colors = []
         elements = []
         i = 0
 
         for vertex in self.mesh.vertices():
-            assert isinstance(vertex, int)
             positions.append(self.mesh.vertex_coordinates(vertex))
-            colors.append(self.pointscolor.get(vertex, self.pointscolor["_default"]))  # type: ignore
+            colors.append(self.vertexcolor[vertex] or self.vertexcolor.default)  # type: ignore
```
Comment (tomvanmele) [inline src/compas_viewer/scene/meshobject.py:103]: `or self.vertexcolor.default` is not necessary. that is the default behaviour...

File: src/compas_viewer/scene/meshobject.py [hunk 8]
```diff
@@ -42,42 +51,68 @@ class MeshObject(ViewerSceneObject, BaseMeshObject):
...
-    def _read_lines_data(self) -> DataType:
+    def _read_lines_data(self) -> ShaderDataType:
         positions = []
         colors = []
         elements = []
         i = 0
 
         for u, v in self.mesh.edges():
-            color = self.linescolor.get((u, v), self.linescolor["_default"])  # type: ignore
+            color = self.edgecolor[(u, v)] or self.edgecolor.default  # type: ignore
```
Comment (tomvanmele) [inline src/compas_viewer/scene/meshobject.py:115]: `or self.edgecolor.default` is not necessary. that is the default behaviour...

File: src/compas_viewer/scene/meshobject.py [hunk 9]
```diff
@@ -97,25 +132,24 @@ def _read_lines_data(self) -> DataType:
...
             if len(vertices) == 3:
                 a, b, c = vertices
                 positions.append(self.mesh.vertex_coordinates(a))
                 positions.append(self.mesh.vertex_coordinates(b))
                 positions.append(self.mesh.vertex_coordinates(c))
                 if self.use_vertexcolors:
-                    colors.append(self.vertexcolor[a])
-                    colors.append(self.vertexcolor[b])
-                    colors.append(self.vertexcolor[c])
+                    colors.append(self.vertexcolor[a] or self.vertexcolor.default)  # type: ignore
```
Comment (tomvanmele) [inline src/compas_viewer/scene/meshobject.py:150]: same comment about `or default`

File: src/compas_viewer/scene/meshobject.py [hunk 10]
```diff
@@ -131,12 +165,12 @@ def _read_frontfaces_data(self) -> DataType:
                 positions.append(self.mesh.vertex_coordinates(c))
                 positions.append(self.mesh.vertex_coordinates(d))
                 if self.use_vertexcolors:
-                    colors.append(self.vertexcolor[a])
-                    colors.append(self.vertexcolor[b])
-                    colors.append(self.vertexcolor[c])
-                    colors.append(self.vertexcolor[a])
-                    colors.append(self.vertexcolor[c])
-                    colors.append(self.vertexcolor[d])
+                    colors.append(self.vertexcolor[a] or self.vertexcolor.default)  # type: ignore
```
Comment (tomvanmele) [inline src/compas_viewer/scene/meshobject.py:168]: everywhere, will not repeat it...

File: src/compas_viewer/scene/nurbssurfaceobject.py [hunk 11]
```diff
@@ -17,52 +19,36 @@ class NurbsSurfaceObject(ViewerSceneObject, GeometryObject):
...
-        self.v = kwargs.get("v", int(16 + (0 * self.LINEARDEFLECTION)))
-
-        self._triangles = [list(point) for triangle in surface.to_triangles(nu=self.u, nv=self.v) for point in triangle]
-
-    def _read_points_data(self) -> DataType:
-        positions = [Point(*pt) for pt in flatten(self.geometry.points)]
-        colors = [self.pointscolor.get(i, self.pointscolor["_default"]) for i in range(len(positions))]  # type: ignore
-        elements = [[i] for i in range(len(positions))]
-        return positions, colors, elements
+        self.u = int(16 + (0 * self.LINEARDEFLECTION))
```
Comment (tomvanmele) [inline src/compas_viewer/scene/nurbssurfaceobject.py:25]: `int(16 + (0 * self.LINEARDEFLECTION))` is always 16. not sure what you meant here

Comment (tomvanmele) [reply src/compas_viewer/scene/nurbssurfaceobject.py:25]: please use `compas.tolerance.TOL.lineardeflection` instead. the user has otherwise no control over this...

File: src/compas_viewer/scene/nurbssurfaceobject.py [hunk 12]
```diff
@@ -17,52 +19,36 @@ class NurbsSurfaceObject(ViewerSceneObject, GeometryObject):
...
 
-    def _read_lines_data(self) -> DataType:
-        positions = [Point(*pt) for pt in flatten(self.geometry.points)]
-        colors = [self.linescolor.get(i, self.linescolor["_default"]) for i in range(len(positions))]  # type: ignore
-        count = 0
+    @property
+    def points(self) -> Optional[list[Point]]:
+        """The points to be shown in the viewer."""
+        points = []
+        for row in self.geometry.points:
```
Comment (tomvanmele) [inline src/compas_viewer/scene/nurbssurfaceobject.py:32]: for a nurbssurface, these points are the control points...

File: src/compas_viewer/scene/nurbssurfaceobject.py [hunk 13]
```diff
@@ -17,52 +19,36 @@ class NurbsSurfaceObject(ViewerSceneObject, GeometryObject):
...
+            points.extend(row)
+        return points
 
-        indexes = []
+    @property
+    def lines(self) -> Optional[list[Line]]:
```
Comment (tomvanmele) [inline src/compas_viewer/scene/nurbssurfaceobject.py:37]: this will result in visualisation of the control polyhedron of the surface. not sure if that is what is meant to happen here...

File: src/compas_viewer/scene/pointobject.py [hunk 14]
```diff
@@ -21,17 +24,21 @@ class PointObject(ViewerSceneObject, GeometryObject):
...
-        return positions, colors, elements
+    @property
+    def points(self) -> Optional[list[Point]]:
+        """The points to be shown in the viewer."""
+        return [self.geometry]
+
+    @property
+    def lines(self) -> Optional[list[Line]]:
+        """The lines to be shown in the viewer."""
+        pass
```
Comment (tomvanmele) [inline src/compas_viewer/scene/pointobject.py:39]: i think this should be an explicit `return None`

File: src/compas_viewer/scene/pointobject.py [hunk 15]
```diff
@@ -21,17 +24,21 @@ class PointObject(ViewerSceneObject, GeometryObject):
...
 
-    def _read_lines_data(self):
-        """No line data exist for this geometry, Return None."""
-        return None
+    @property
+    def surfaces(self) -> Optional[list[Tuple[Point, Point, Point]]]:
+        """The surface to be shown in the viewer. Currently only triangles are supported."""
+        pass
```
Comment (tomvanmele) [inline src/compas_viewer/scene/pointobject.py:44]: same here...

File: src/compas_viewer/scene/polygonobject.py [hunk 16]
```diff
@@ -13,4 +18,20 @@ class PolygonObject(MeshObject):
...
+        return self.geometry.points
+
+    @property
+    def lines(self) -> Optional[list[Line]]:
+        return self.geometry.lines
+
+    @property
+    def surfaces(self) -> Optional[list[Tuple[Point, Point, Point]]]:
+        """The surface to be shown in the viewer. Currently only triangles are supported."""
+        pass
```
Comment (tomvanmele) [inline src/compas_viewer/scene/polygonobject.py:37]: same everywhere. will not repeat it...

File: src/compas_viewer/scene/scene.py [hunk 17]
```diff
@@ -142,28 +133,18 @@ def add(
...
+            linecolor=linecolor,
+            facecolor=facecolor,
+            surfacecolor=surfacecolor,
+            vertexcolor=vertexcolor,
+            edgecolor=edgecolor,
             lineswidth=lineswidth,
             pointssize=pointssize,
             opacity=opacity,
             hide_coplanaredges=hide_coplanaredges,
             use_vertexcolors=use_vertexcolors,
```
Comment (tomvanmele) [inline src/compas_viewer/scene/scene.py:146]: why are all these settings moved here?

Comment (tomvanmele) [reply src/compas_viewer/scene/scene.py:146]: they belong to the scene object...

File: src/compas_viewer/scene/sceneobject.py [hunk 18]
```diff
@@ -13,18 +12,18 @@
 from numpy import average
 from numpy import identity
 
-from compas_viewer.configurations import SceneConfig
+from compas_viewer.components.renderer.shaders import Shader
 from compas_viewer.utilities.gl import make_index_buffer
```
Comment (tomvanmele) [inline src/compas_viewer/scene/sceneobject.py:16]: not a big fan of `utilities`. why not just `compas_viewer.gl`?

File: src/compas_viewer/components/renderer/renderer.py [hunk 19]
```diff
@@ -362,15 +362,15 @@ def init(self):
         """Initialize the renderer."""
 
         # Init the grid
-        self.grid = self.scene.add(  # type: ignore
-            Frame.worldXY(),
+        self.grid: FrameObject = self.scene.add(  # type: ignore
```
[the remark this answers is not in the document]
Comment (ZacZhangzhuo) [reply src/compas_viewer/components/renderer/renderer.py:365]: sorry forgot this... fixed

File: src/compas_viewer/scene/geometryobject.py [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+        The triangulated mesh representation of the geometry.
+    LINEARDEFLECTION : float
+        The default linear deflection for the geometry.
+
+    See Also
+    --------
+    :class:`compas.geometry.Geometry`
+    """
+
+    LINEARDEFLECTION = 0.2
```
[the remark this answers is not in the document]
Comment (ZacZhangzhuo) [reply src/compas_viewer/scene/geometryobject.py:51]: Yes, we talked about this and this was the temporary solution at the time. The problem is, currently `compas.tolerance.TOL.lineardeflection` =1e-3 which is too small and super detailed. should I modify the number?

File: src/compas_viewer/scene/geometryobject.py (hunk 2 above)
Comment (ZacZhangzhuo) [reply src/compas_viewer/scene/geometryobject.py:83]: changed `surfaces` to `viewmesh`

Comment (tomvanmele) [reply src/compas_viewer/scene/geometryobject.py:83]: that is not really what i meant. many geometry objects use a mesh representation for viewing purposes. currently, it seems that the construction of this "viewmesh" is more or less repeated for many objects. perhaps this coud be generalised with a `viewmesh` property...

File: src/compas_viewer/scene/geometryobject.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tomvanmele) [reply src/compas_viewer/scene/geometryobject.py:51]: i think you are using this value in the wrong way. `u` and `v` parameters for the discretisation of `compas.geometry` shapes should be configurable by the user, with a sensible default like `u=64` and `v=u`. linear deflection is used when discretising Nurbs and Brep. there it depends on the size of the model. since most people tend to work in meters, 1e-3 makes sense. if they work in mm they can set the linear deflection to 1 or 0.1 ...

File: src/compas_viewer/scene/nurbssurfaceobject.py (hunk 12 above)
Comment (ZacZhangzhuo) [reply src/compas_viewer/scene/nurbssurfaceobject.py:32]: I  think we want to display the control points no?

File: src/compas_viewer/scene/nurbssurfaceobject.py (hunk 13 above)
Comment (ZacZhangzhuo) [reply src/compas_viewer/scene/nurbssurfaceobject.py:37]: image

File: src/compas_viewer/scene/meshobject.py (hunk 7 above)
Comment (ZacZhangzhuo) [reply src/compas_viewer/scene/meshobject.py:103]: That did not work when i tried. sometimes `vertexcolor[vertex]` does exist but is `None`, thus returns a `None`
